### ISSUE TITLE
About Dataparallel

### ISSUE THREAD

**Issue by emjay73** [dbolya/yolact#238]
About all the hacky things that you wrote with comments, especially the things that have something to do with '(custom)dataparallel', is there anything that I have to know about its property when I want to add or remove some layers to the code? (multi GPU usage case)

Could you explain the following lines in detail for me ?

case1)
```
//A hack for the way dataparallel works
if cfg.share_prediction_module and pred_layer is not self.prediction_layers[0]:
    pred_layer.parent = [self.prediction_layers[0]]
=> parent double assignment on purpose?
```
(assignment already happened in PredictionModule constructor)

case2)
```
elif self.priors.device != device:
                # This whole weird situation is so that DataParalell doesn't copy the priors each iteration
                if prior_cache[size] is None:
                    prior_cache[size] = {}
                
                if device not in prior_cache[size]:
                    prior_cache[size][device] = self.priors.to(device)

                self.priors = prior_cache[size][device]
```

=> (whaaaat..?) device id consideration?

Well, I want to modify the code and retrain it with multiple GPUs but I'm afraid I could mess things up because of my lack of knowledge.

**Comment by dbolya** [comment]
1. I think actually that comment was for a line of code I deleted earlier, whoops. The hack is actually for the way modules work, not dataparallel: modules count any module that's a member variable as part if its parameters and state_dict. However, I want to share the same "parent" module between multiple prediction heads (since FPN uses a shared prediction head). So to get around module's code, I cheekily put the parent in a python array which makes Module think it's not another module and thus doesn't add it to its state_dict. I don't think change is necessary there for additional layers, but keep that explanation in mind.

2. So the way dataparallel works is that it replicates the model on every GPU, computes the loss, then destroys the replicated copies. This makes sense if the things it's copying are weights that change for instance from one iteration to the next, but for priors which are fixed that's super inefficient. So I just have a cache for priors on each device so we don't copy them every iteration (also cross-device computation is a big no-no so we can't just have one copy on GPU 0).

**Comment by emjay73** [comment]
1. To recap, a 'parent' should be wrapped with [] in order to share parameters between multiple prediction heads..?

One related, additional question. 
Why do you assign parents two times?
1) PredictionModule constructor
 **yolact __init__**
for idx, layer_idx in enumerate(self.selected_layers):
            # If we're sharing prediction module weights, have every module's parent be the first one
            parent = None
            if cfg.share_prediction_module and idx > 0:
                parent = self.prediction_layers[0]      **<----- here**

            pred = PredictionModule(src_channels[layer_idx], src_channels[layer_idx],
                                    aspect_ratios = cfg.backbone.pred_aspect_ratios[idx],
                                    scales        = cfg.backbone.pred_scales[idx],
                                    parent        = parent,   **<----- here**
                                    index         = idx)
            self.prediction_layers.append(pred)


**PredictionModule __init__**
self.parent      = [parent] # Don't include this in the state dict **<----- here**

2) yolact forward
for idx, pred_layer in zip(self.selected_layers, self.prediction_layers):
                pred_x = outs[idx]

                if cfg.mask_type == mask_type.lincomb and cfg.mask_proto_prototypes_as_features:
                    # Scale the prototypes down to the current prediction layer's size and add it as inputs
                    proto_downsampled = F.interpolate(proto_downsampled, size=outs[idx].size()[2:], mode='bilinear', align_corners=False)
                    pred_x = torch.cat([pred_x, proto_downsampled], dim=1)

                # A hack for the way dataparallel works
                if cfg.share_prediction_module and pred_layer is not self.prediction_layers[0]:
                    pred_layer.parent = [self.prediction_layers[0]] **<----- here**

                p = pred_layer(pred_x)
                
                for k, v in p.items():
                    pred_outs[k].append(v)

2.  Oh... That's the reason why you used the global variable 'prior_cache' and save it for each device... I see..

**Comment by dbolya** [comment]
1. Yup

---

1. Oh... I guess that's an artifact of when I didn't use CustomDataparallel. I'll try removing the assignment at (2) to see what it does.

**Comment by dbolya** [comment]
Ok it looks like assigning it every iteration is necessary, because if we don't then the parent wont be on the same device as you're executing on. So it's really the first assignment that's unnecessary, because you definitely have to assign it every iteration. I think there's not much reason to change it now though.

**Comment by emjay73** [comment]
I get it.
Thank you for your detailed explanation.
I learn many things from you!

**Comment by dbolya** [comment]
No problem! I'll close the issue now.
